File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/com/ririnto/sinon/harness/rules/text/HookStageRule.java
package com.ririnto.sinon.harness.rules.text;

import com.ririnto.sinon.harness.rules.HarnessCheckHelper;
import com.ririnto.sinon.harness.rules.HarnessCheckRule;
import com.ririnto.sinon.harness.core.RuleContext;
import com.ririnto.sinon.harness.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;
import java.util.stream.Stream;

/**
 * Rule that requires hook files to contain stage markers.
 */
public enum HookStageRule implements HarnessCheckRule {
    INSTANCE;

    @Override
    public String category() {
        return "hookStage";
    }
    private static final String CATEGORY = "hookStage";

    @Override
    public boolean applies(RuleContext ctx) {
        return ctx.manifest().isEnabled(CATEGORY);
    }

    @Override
    public Collection<Finding> validate(RuleContext ctx) throws MojoExecutionException {
        final Path root = ctx.root();
        final JsonNode manifest = ctx.manifest().raw();
        final JsonNode catNode = manifest.get(CATEGORY);
        final JsonNode stages = catNode.get("parameters").get("stages");
        final JsonNode stackStages = stages.get("maven");
        final String markerTemplate = catNode.get("parameters").get("markerTemplate").asText();
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return Stream.ofNullable(stackStages)
                .flatMap(ss -> ss.propertyNames().stream()
                        .flatMap(hookName -> validateStage(root, hookName, ss.get(hookName).asText(), markerTemplate, severity).stream()))
                .toList();
    }

    private List<Finding> validateStage(Path root, String hookName, String expectedStage, String markerTemplate, String severity) {
        final String hookPath = "docs/harness/git-hooks/" + hookName;
        final Path hook = root.resolve(hookPath);
        return Stream.of(hook)
                .filter(p -> HarnessCheckHelper.isSafeRegularFile(root, p))
                .flatMap(p -> {
                    try {
                        final String text = HarnessCheckHelper.readFile(root, p);
                        final String expectedMarker = markerTemplate.replace("{stage}", expectedStage);
                        return text.contains(expectedMarker)
                                ? Stream.empty()
                                : Stream.of(Finding.of(severity, CATEGORY, hookPath + " must contain stage marker '" + expectedMarker + "'"));
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/com/ririnto/sinon/harness/rules/text/HookStageRule.kt
package com.ririnto.sinon.harness.rules.text

import com.ririnto.sinon.harness.ast.HarnessAstResults.Finding
import com.ririnto.sinon.harness.core.JsonAccess
import com.ririnto.sinon.harness.core.RuleContext
import com.ririnto.sinon.harness.rules.HarnessCheckRule
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isRegularFile
import kotlin.io.path.readText

/**
 * Rule that requires hooks to contain stage markers.
 */
object HookStageRule : HarnessCheckRule() {
    /**
     * Category key.
     */
    override val category: String = "hookStage"

    override fun validate(ctx: RuleContext): Collection<Finding> =
        buildList {
            val catObj = ctx.manifest.categoryObject(category)
            val parametersObj = catObj?.get("parameters")?.jsonObject
            val stagesObj = parametersObj?.get("stages")?.jsonObject
            if (catObj != null && parametersObj != null && stagesObj != null) {
                val markerTemplate = JsonAccess.stringFromObject(parametersObj, "markerTemplate")
                val preCommitStage = JsonAccess.stringFromObject(stagesObj, "pre-commit")
                val prePushStage = JsonAccess.stringFromObject(stagesObj, "pre-push")
                val preCommitHook = ctx.root / "docs/harness/git-hooks/pre-commit"
                val prePushHook = ctx.root / "docs/harness/git-hooks/pre-push"
                if (preCommitHook.isRegularFile()) {
                    if (!preCommitHook.readText().contains(markerTemplate.replace("{stage}", preCommitStage))) {
                        add(
                            Finding(
                                ctx.manifest.severityOf(category),
                                category,
                                ctx.manifest.stringValue(category, "default").takeIf { message ->
                                    message.isNotEmpty()
                                }
                                    ?: "pre-commit must contain stage marker '${markerTemplate.replace(
                                        "{stage}",
                                        preCommitStage,
                                    )}'",
                            ),
                        )
                    }
                }
                if (prePushHook.isRegularFile()) {
                    if (!prePushHook.readText().contains(markerTemplate.replace("{stage}", prePushStage))) {
                        add(
                            Finding(
                                ctx.manifest.severityOf(category),
                                category,
                                ctx.manifest.stringValue(category, "default").takeIf { message ->
                                    message.isNotEmpty()
                                }
                                    ?: "pre-push must contain stage marker '${markerTemplate.replace(
                                        "{stage}",
                                        prePushStage,
                                    )}'",
                            ),
                        )
                    }
                }
            }
        }
}
